saleor/payment/gateways/adyen/webhooks.py

```diff
@@ -616,10 +623,113 @@ def handle_order_opened(notification: Dict[str, Any], gateway_config: GatewayCon
+    handle partial payments on their side and send us info about them. We want to
+    somehow store some basic information about this but without modifying a whole
+    Saleor logic. We're going to change it by introducing partial payments feature on
+    Saleor side.
+    """
+    additional_data = notification.get("additionalData", {})
+    new_payments = []
+
+    for (
+        payment_method_key,
+        payment_amount_key,
+        psp_reference_key,
+    ) in get_adyen_partial_payment_keys(additional_data):
+        # payment amount key has structure - "{currency} {amount}", like 'GBP 41.90'
+        currency, amount = additional_data[payment_amount_key].split()
+
+        new_payment = Payment(**model_to_dict(payment, exclude=["id", "checkout"]))
+        new_payment.checkout_id = payment.checkout_id
+        new_payment.is_active = False
+        new_payment.partial = True
+        new_payment.total = Decimal(amount)
+        # Increasing captured amount increases a total paid of the order.
+        new_payment.captured_amount = 0
+        new_payment.currency = currency
+        new_payment.payment_method_type = additional_data[payment_method_key]
+        new_payment.psp_reference = additional_data[psp_reference_key]
+        new_payment.extra_data = json.dumps({"parent_payment_id": payment.id})
+        new_payments.append(new_payment)
+
+    already_existing_partial_payments = Payment.objects.filter(
+        psp_reference__in=[p.psp_reference for p in new_payments],
+        partial=True,
+        checkout_id=payment.checkout_id,
+    )
+    if already_existing_partial_payments:
+        return list(already_existing_partial_payments)
+
+    if new_payments:
+        return Payment.objects.bulk_create(new_payments)
+    return None
+
+
+def create_order_event_about_adyen_partial_payments(
+    notification: Dict[str, Any], payment
+):
+    additional_data = notification.get("additionalData", {})
+    msg = "Adyen: "
+
+    for (
+        payment_method_key,
+        payment_amount_key,
+        psp_reference_key,
+    ) in get_adyen_partial_payment_keys(additional_data):
+        payment_method = additional_data[payment_method_key]
+        payment_amount = additional_data[payment_amount_key]
+        psp_reference = additional_data[psp_reference_key]
+        msg += (
+            f"Partial payment with {payment_method} (PSP: {psp_reference}) "
+            f"for {payment_amount}.\n"
+        )
+    create_payment_notification_for_order(
+        payment, msg, failed_msg=None, is_success=True
+    )
+
+
+def get_adyen_partial_payment_keys(additiona_data):
+    """Get keys that contains adyen partial data.
+
+    The data received from Adyen has strange structure. In middle of dict key, there is
+    a int number pointing to the given payment method. So we need to convert it and
+    match somehow a paymentMethod, paymentAmount and payment psp reference
+    """
+    for i in range(1, 99):
```

Comment:
I'm wondering about this range here and it's values ?

Reply:
🤦  right...